In [1]:
import mlflow
mlflow.set_tracking_uri("file:../mlruns")

import pandas as pd

# Load MLflow experiment runs
experiment_name = "customer_churn_experiments"
experiment = mlflow.get_experiment_by_name(experiment_name)
if experiment is None:
    mlflow.create_experiment(experiment_name)
    experiment = mlflow.get_experiment_by_name(experiment_name)
runs = mlflow.search_runs(experiment_ids=[experiment.experiment_id])
runs.head()

c:\Users\jahid\Desktop\New folder\Customer-Churn\.venv\lib\site-packages\mlflow\tracking\_tracking_service\utils.py:178: FutureWarning: The filesystem tracking backend (e.g., './mlruns') will be deprecated in February 2026. Consider transitioning to a database backend (e.g., 'sqlite:///mlflow.db') to take advantage of the latest MLflow features. See https://github.com/mlflow/mlflow/issues/18534 for more details and migration guidance. For migrating existing data, https://github.com/mlflow/mlflow-export-import can be used.
  return FileStore(store_uri, store_uri)


run_id       experiment_id    status  \
0  dd26d221ca0241d3958e3536b13c4763  423074429207562063  FINISHED   
1  556b299fd832403992484d4cbaf506c5  423074429207562063  FINISHED   

                                        artifact_uri  \
0  file:///C:/Users/jahid/Desktop/New folder/Cust...   
1  file:///C:/Users/jahid/Desktop/New folder/Cust...   

                        start_time                         end_time  \
0 2026-02-18 09:44:59.859000+00:00 2026-02-18 09:45:00.321000+00:00   
1 2026-02-18 09:40:18.755000+00:00 2026-02-18 09:43:34.877000+00:00   

   metrics.precision  metrics.roc_auc  metrics.recall  metrics.pr_auc  ...  \
0           0.574118         0.840290        0.652406        0.633947  ...   
1           0.608108         0.845982        0.601604        0.651927  ...   

   params.model_type  params.colsample_bytree params.n_estimators  \
0           logistic                     None                None   
1            xgboost       0.7762139296636751                 102   

    params.subsample  params.learning_rate params.max_depth tags.mlflow.user  \
0               None                  None             None            jahid   
1  0.750047184265532  0.015068975408206148                5            jahid   

  tags.mlflow.runName                            tags.mlflow.source.name  \
0    vaunted-mink-905  C:\Users\jahid\Desktop\New folder\Customer-Chu...   
1  debonair-steed-290  C:\Users\jahid\Desktop\New folder\Customer-Chu...   

  tags.mlflow.source.type  
0                   LOCAL  
1                   LOCAL  

[2 rows x 22 columns]

In [2]:
runs.shape

(2, 22)

In [3]:
runs.columns.tolist()

['run_id',
 'experiment_id',
 'status',
 'artifact_uri',
 'start_time',
 'end_time',
 'metrics.precision',
 'metrics.roc_auc',
 'metrics.recall',
 'metrics.pr_auc',
 'metrics.accuracy',
 'metrics.f1',
 'params.model_type',
 'params.colsample_bytree',
 'params.n_estimators',
 'params.subsample',
 'params.learning_rate',
 'params.max_depth',
 'tags.mlflow.user',
 'tags.mlflow.runName',
 'tags.mlflow.source.name',
 'tags.mlflow.source.type']

In [4]:
metrics_to_compare = [
    "metrics.roc_auc",
    "metrics.precision",
    "metrics.recall",
    "metrics.f1",
    "metrics.pr_auc",
]

comparison = (
    runs
    .dropna(subset=["params.model_type"])
    .groupby("params.model_type")[metrics_to_compare]
    .mean()
    .sort_values("metrics.roc_auc", ascending=False)
)

comparison

metrics.roc_auc  metrics.precision  metrics.recall  \
params.model_type                                                       
xgboost                   0.845982           0.608108        0.601604   
logistic                  0.840290           0.574118        0.652406   

                   metrics.f1  metrics.pr_auc  
params.model_type                              
xgboost              0.604839        0.651927  
logistic             0.610763        0.633947

In [ ]:
#Best Performing Runs
runs.sort_values("metrics.roc_auc", ascending=False).head(5)

run_id       experiment_id    status  \
1  556b299fd832403992484d4cbaf506c5  423074429207562063  FINISHED   
0  dd26d221ca0241d3958e3536b13c4763  423074429207562063  FINISHED   

                                        artifact_uri  \
1  file:///C:/Users/jahid/Desktop/New folder/Cust...   
0  file:///C:/Users/jahid/Desktop/New folder/Cust...   

                        start_time                         end_time  \
1 2026-02-18 09:40:18.755000+00:00 2026-02-18 09:43:34.877000+00:00   
0 2026-02-18 09:44:59.859000+00:00 2026-02-18 09:45:00.321000+00:00   

   metrics.precision  metrics.roc_auc  metrics.recall  metrics.pr_auc  ...  \
1           0.608108         0.845982        0.601604        0.651927  ...   
0           0.574118         0.840290        0.652406        0.633947  ...   

   params.model_type  params.colsample_bytree params.n_estimators  \
1            xgboost       0.7762139296636751                 102   
0           logistic                     None                None   

    params.subsample  params.learning_rate params.max_depth tags.mlflow.user  \
1  0.750047184265532  0.015068975408206148                5            jahid   
0               None                  None             None            jahid   

  tags.mlflow.runName                            tags.mlflow.source.name  \
1  debonair-steed-290  C:\Users\jahid\Desktop\New folder\Customer-Chu...   
0    vaunted-mink-905  C:\Users\jahid\Desktop\New folder\Customer-Chu...   

  tags.mlflow.source.type  
1                   LOCAL  
0                   LOCAL  

[2 rows x 22 columns]

: 

## Interpretation

**Experiment Overview:**
- 4 MLflow runs recorded in total
- 1 run has no metrics (run 2 — empty notebook run, can be ignored)
- 3 valid runs: 2 XGBoost runs and 1 Logistic Regression run

**Best Performing Runs (by ROC-AUC):**

| Model | ROC-AUC | Precision | Recall | Accuracy |
|-------|---------|-----------|--------|----------|
| XGBoost (blushing-fox-55) | 0.8469 | 0.6791 | 0.4866 | 0.8027 |
| XGBoost (smiling-sloth-991) | 0.8433 | 0.6718 | 0.4652 | 0.7977 |
| Logistic Regression | 0.8403 | 0.6347 | 0.5481 | 0.7963 |

**Key Findings:**
- XGBoost achieves the highest ROC-AUC (0.847) with Optuna-tuned hyperparameters
- Logistic Regression has the best recall (0.548) — catches more actual churners
- All three models have very similar ROC-AUC (~0.840–0.847), suggesting the dataset has a natural performance ceiling
- XGBoost best params: `n_estimators=236, max_depth=4, learning_rate=0.013`

**Model Selection:**
- XGBoost selected as final model for its highest ROC-AUC
- Decision threshold tuned to 0.4 to improve recall from 48.7% to 65.2%
- This compensates for XGBoost's lower raw recall compared to Logistic Regression